### DISCUSSION TITLE
StateNotifier, listen Streams or use StreamProvider?

### DISCUSSION THREAD

**Discussion by todoFixIt** [rrousselGit/riverpod#1490] [Q&A] [answered]
I have a project with a DB, DAOs, and Services that expose some of the DAOs' methods. Several of those methods return `Stream`. By architecture design I choose to have a **controller** (`StateNotifier`) per **screen** and the screen widget only have access to the `Provider` of that **controller**.

The **controller** will expose some of the **service's** methods, so I need the **service** as a parameter, but I also need the values from the `Stream` as part of the **controler**'s **state**. I've seen both in the docs and in another examples the `StreamProvider` but it "felt" wrong to have a `StreamProvider` per `Stream` of the **service** since the **controller** already has and needs the **service** and its the one with the `Stream`.

Is it a bad practice or are there any issues (memory leaks...) to listen to the `Stream` inside the **controller's** constructor or should I create a `StreamProvider` for each `Stream` I have on the **service**? 

An example of what I mean:
```dart
class ClientService {
  ...
  Stream<List<Client>> watchAllClients() => _dao.watchAllClients();
  
  void insertClient(Client client) => _dao.insertClient();
}

final clientServiceProvider = Provider<ClientService>((ref) {
  return ClientService(...);
});

class MyState {
  final List<Client> clients;
  ...
}

class MyController extends StateController<MyState> {
  final ClientService _clientService;
  ...
  MyController(this._clientService, ...) : super(MyState(...)) {
    _listenToClients();
  }
  
  void insertClient(Client client) => _clientService.insertClient();

  void _listenToClients() => _clientService.watchAllClients().listen((e) {
      state = state.copyWith(clients: e);
  });
}

final myControllerProvider = StateNotifierProvider.autoDispose<MyController, MyState>((ref) {
  ...
  final service = ref.watch(clientServiceProvider);
  return MyController(service,...);
});
```

**Comment by carloshpb** [comment]
@Gilga89 I'm having right now the same question in my head. From what I've been researching (for some hours), I think it'll be the same result (probably).

The StreamProvider would be an easier and more direct way to get the actual state from the Stream and dispose of it.

While the way you did in the code would be good if you're following an architecture (i.e. clean) and you're planning to keep every VM in the same way, so as to not confuse the other developers.

But still .... I got the same feelings that you have at the moment lol.
I hope that someone from the team can answer this one to take off some doubts.

**Comment by rrousselGit** [comment, marked as answer]
You're free to manually listen to anything

You'll have to make sure that you cancel the stream subscription though. Your snippet currently doesn't.

Maybe consider:

```dart
final sub = stream.listen(...);
ref.onDispose(sub.cancel);
```
